compiler/rustc_passes/src/dead.rs

```diff
@@ -677,30 +677,91 @@ impl<'tcx> DeadVisitor<'tcx> {
         let inherent_impls = self.tcx.inherent_impls(def_id);
         for &impl_did in inherent_impls.iter() {
             for item_did in self.tcx.associated_item_def_ids(impl_did) {
-                if let Some(def_id) = item_did.as_local() {
-                    if self.live_symbols.contains(&def_id) {
-                        return true;
-                    }
+                if let Some(def_id) = item_did.as_local()
+                    && self.live_symbols.contains(&def_id)
+                {
+                    return true;
                 }
             }
         }
         false
     }
 
-    fn warn_dead_code(
+    fn store_dead_field_or_variant(
         &mut self,
         id: hir::HirId,
         span: rustc_span::Span,
         name: Symbol,
+    ) {
+        let field_or_variant_local_def_id = |parent_node, node| -> Option<(LocalDefId, bool)> {
+            match (parent_node, node) {
+                (
+                    Node::Item(hir::Item {
+                        kind:
+                            hir::ItemKind::Struct(hir::VariantData::Struct(..), ..)
+                            | hir::ItemKind::Union(hir::VariantData::Struct(..), ..),
+                        def_id,
+                        ..
+                    }),
+                    Node::Field(..),
+                ) => Some((*def_id, true)),
+                (
+                    Node::Item(hir::Item { kind: hir::ItemKind::Enum(..), def_id, .. }),
+                    Node::Variant(..),
+                ) => Some((*def_id, false)),
+                (
+                    Node::Variant(hir::Variant { data: hir::VariantData::Struct(..), id, .. }, ..),
+                    Node::Field(..),
+                ) => self
+                    .tcx
+                    .hir()
+                    .find_parent_node(*id)
+                    .and_then(|id| self.tcx.hir().opt_local_def_id(id).map(|id| (id, true))),
+                _ => None,
```

Comment:
Comment here would be nice. My understanding is that first case is struct/union fields, second case is enum variants, and third case is enum variant fields?

Reply:
I added comments.